Status Moves V2

SandytheLizard opened the pull request:
1. A stat that does not need modified does not need to be in the JSON file
2. Instead of divide, it accepts an iterator of +, -, /, *, =
3. you may specify a maxvalue,if you do it randomly picks between the minvalue and maxvalue. value has been replaced by minvalue

SandytheLizard committed:
Status Moves V2

1. A stat that does not need modified does not need to be in the JSON file
2. Instead of divide, it accepts an iterator of +, -, /, *, =
3. you may specify a maxvalue,if you do it randomly picks between the minvalue and maxvalue. value has been replaced by minvalue

vnmabus commented on tuxemon/db.py:
```diff
@@ -106,9 +106,10 @@ class JSONMonster(JSONMonsterOptionalFields):
+class JSONStat(TypedDict, total=False):
+    minvalue: int
+    maxvalue: int
+    iterator: str
```
Why `iterator` instead of `operation`?

SandytheLizard replied:
done

vnmabus commented on tuxemon/db.py:
```diff
@@ -106,9 +106,10 @@ class JSONMonster(JSONMonsterOptionalFields):
-    value: int 
-    dividing: bool
+class JSONStat(TypedDict, total=False):
+    minvalue: int
```
I would prefer `value` and `max_deviation` instead, as it is more clear in the fixed damage case, and it can be generalized to the normal distribution it necessary.

SandytheLizard replied:
fixed

vnmabus commented on mods/tuxemon/db/technique/amnesia.json:
```diff
@@ -3,21 +3,33 @@
   "potency": 1,
   "power": 2.25,
+  "statdodge": {
+    "minvalue": 0, "maxvalue": 3, "iterator":"-", "overridetofull":false
```
Remove `iterator` and `overridetofull` when they have their default values, as you already made them optional. It makes the JSON easier to read.

SandytheLizard replied:
Removed override to full, Iterator does not have default value.

vnmabus commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -171,8 +171,13 @@ msgstr "{user} is leeching health from {target}."
+msgstr "{user} Overfed {target}. {target} 's speed was halved, {target}' s hp was fully restored"
 
+msgid "combat_amnesia_get"
+msgstr "{user} 's dodge fell. \n {user} 's attack damage types rose."
```
Maybe the "attribute x increased/decreased/restored" could be separate strings, so that we only have to translate them once.

SandytheLizard replied:
im not sure i did it right, please take a look

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,38 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
                 continue
-            value = stat['value']
-            dividing = stat['dividing']
+            minvalue = stat['minvalue']
```
As the values of the stat are now optional (because `total=False`), you need to use the `get` method with a default value in case they don't exist. If you configure Mypy in your IDE it should flag this as an error.

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,38 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+            if value > 0 and  override == False:
+                if iterator == '/':
                     newstatvalue = basestatvalue // value
-                success = True
-                setattr(target, slug, newstatvalue)
+                    setattr(target, slug, newstatvalue)
+                elif iterator == '+':
+                    newstatvalue = basestatvalue + value
+                    setattr(target, slug, newstatvalue)
+                elif iterator == '-':
+                    newstatvalue = basestatvalue - value
+                    setattr(target, slug, newstatvalue)
+                elif iterator == '*':
+                    newstatvalue = basestatvalue * value
+                    setattr(target, slug, newstatvalue)
```
This could be done in a more compact way:
```python
ops_dict = {
    "+": operator.add,
    "-": operator.sub,
    "*": operator.mul,
    "/": operator.floordiv,
}

newsatvalue = ops_dict[iterator](basestatvalue, value)
setattr(target, slug, newstatvalue)
```

SandytheLizard replied:
fixed

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,38 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
-                    newstatvalue = basestatvalue + value
-                else:
+            if maxvalue:
+                value = random.randint(minvalue,maxvalue)
```
Using random directly makes everything non deterministic. Maybe we should discuss having something similar to numpy `RandomState` objects, in order to be able to reproduce combats, for networking, debugging or rewatching purposes.

SandytheLizard replied:
going to into a later PR

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,38 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+                setattr(target, slug, newstatvalue)
         return {
-            "success": bool(stat)
+            "success": bool(newstatvalue)
```
I do not understand this success value.

SandytheLizard replied:
i did this because the prorgam returns that an attack failed, even if it hit,  when even one stat is 0. By putting newstatvalue=0 right before the loop, and newstatvalue in the loop once a stat is above 0, it checks for at least one stat being successful rather than all of them.

vnmabus replied:
I still do not get it. What happens if an attack reduces a stat to 0?

SandytheLizard replied:
then it fails because a stat cant be 0

vnmabus replied:
What about hp?

bitcraft replied:
yeah, current_hp is a problem.  if it is set to 0, then it is a "success".  you should check if there is a change from the old stat to the new one, then check for success.  as is, if a stat change KOs a monster, then the dialog will say it failed.

SandytheLizard replied:
fixed by checking if the stat is hp and if so the newstatvalue is equal to 1 which is always true

bitcraft replied:
what about the other stats?  why are you so sure they will never be 0?

SandytheLizard replied:
im honestly unsure why i thought that, i added a check that makes the value always at least 1 before it returns the value

bitcraft replied:
ok, just calrifing then, you are aware that this will never be false?  i see that `statdodge` has a min of 0 in the JSON.  consider if the dodge value is 3 at the start, then we get 0 for the diff.  the final result will be 3, which your function will call a success.  is that what you intend to do?  ideally, you would take the difference of the old and new, so you can reliably tell if the technique did anything or not.  because as-is, it ill always register a success, even if there was no change to the stats.

SandytheLizard replied:
i think itll be fine as-is, as unless the move is coded badly and somehow affects stats but has no stat changes, theres no reason as to why the stats wouldn't change

SandytheLizard committed:
Update amnesia.json

Removed override to full, Iterator does not have default value.

SandytheLizard committed:
Update amnesia.json

Changed Iterator to Operator

SandytheLizard committed:
Update db.py

Changed to Value and Max Deviation

SandytheLizard committed:
Update db.py

Changed Iterator to Operation

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update base.po

Added some new lines, not sure if this is the right way to do it

SandytheLizard committed:
Update base.po

SandytheLizard committed:
Update base.po

SandytheLizard committed:
Update technique.py

vnmabus commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,30 @@ msgstr "{name} took poison damage!"
 msgstr "{user} is leeching health from {target}."
 
+msgid "dodge_increased"
+msgstr {user} 's dodge increasedd"
```
I hope the messages can be made generic:
```suggestion
msgstr {user} 's {stat} increased"
```

vnmabus commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,30 @@ msgstr "{name} took poison damage!"
+msgstr {user} 's hp was restored"
+
 msgid "combat_state_overfeed_get"
+msgstr "{user} Overfed {target}" + speed_halved + hp_restored
```
Is this syntax even legal?

SandytheLizard replied:
no, i was still testing stuff and didnt mean to update that yet. its fixed now,

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,32 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+                value = minvalue
+            if value > 0 and  override == False:
+                ops_dict = {
+                "+": operator.add,
```
You need to import the operator module.

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,32 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+                "-": operator.sub,
+                "*": operator.mul,
+                "/": operator.floordiv,
+            }
```
Bad identation.

SandytheLizard replied:
should be fixed

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,32 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
-            value = stat['value']
-            dividing = stat['dividing']
+            minvalue = stat['minvalue']
+            maxvalue = stat['maxvalue']
```
You need to compute them from `value` and `max_deviation`.

SandytheLizard committed:
Update mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po

Co-authored-by: Carlos Ramos Carreño <carlosramosca@hotmail.com>

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +297,32 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+                "/": operator.floordiv,
+            }
+
+                newstatvalue = ops_dict[operator](basestatvalue, value)
```
I have just noticed that there is a name clash between the variable `operator` and the module of the same name. The variable should be `operation` instead.

SandytheLizard replied:
fixed

SandytheLizard committed:
Update technique.py

Fixed the Variable Names, also includes some code to revert stat changes after a battle, wont do anything until the next PR

SandytheLizard committed:
Update base.po

FIxed Invalid Syntax

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
 msgstr "{user} is leeching health from {target}."
 
+msgid "dodge_increased"
+msgstr {user} 's {stat} increased"
```
why is there a space between the name and the `'s`?

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr {user} 's speed was halved"
+
+msgid "hp_restored"
+msgstr {user} 's hp was restored"
```
missing `.` or `!`

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr {user} 's {stat} increased"
+
+msgid "speed_halved"
+msgstr {user} 's speed was halved"
```
punctuation

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr "{user} rose higher into the air."
 
+msgid "combat_amnesia_get"
+msgstr "{user} has gotten amnesia"
```
punctuation

bitcraft commented on tuxemon/technique.py:
```diff
@@ -214,7 +215,11 @@ def full_recharge(self) -> None:
         self._combat_counter = 0
-
+    def keep_old_stats(self):
+        for player in self.active_players:
```
technically, there could be more than oner player, or players could leave.  maybe not now, but in the future, maybe.  this is going to cause problems.  old_stats should keep a reference to the monster.  this should be a dictionary

SandytheLizard committed:
Update base.po

Punctuation and Spacing

SandytheLizard committed:
Update technique.py

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +302,34 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
-                target.current_hp = target.hp
+            value = stat.get('value')
+            max_deviation = stat.get('max_deviation')
+            operation = stat.get('operation')
```
`operation` and `overridetofull` should have default values different from `None`.

vnmabus replied:
It is easier than that: you just have to pass the default value as the second parameter of `get`.

SandytheLizard committed:
Update technique.py

vnmabus commented on tuxemon/technique.py:
```diff
@@ -297,25 +302,34 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+            override = stat.get('overridetofull')
+            basestatvalue = getattr(target, slugdata)
+            if max_deviation:
+                value = random.randint(value,max_deviation)
```
The `max_deviation` is the maximum deviation from the mean value.
```suggestion
                value = random.randint(value - max_deviation, value + max_deviation)
```

SandytheLizard committed:
Update overfeed.json

SandytheLizard committed:
Update amnesia.json

SandytheLizard committed:
Update tuxemon/technique.py

Co-authored-by: Carlos Ramos Carreño <carlosramosca@hotmail.com>

SandytheLizard committed:
Update levitate.json

SandytheLizard committed:
Update levitate.json

SandytheLizard committed:
Update amnesia.json

SandytheLizard committed:
Update levitate.json

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update technique.py

bitcraft commented on mods/tuxemon/db/technique/amnesia.json:
```diff
@@ -3,21 +3,33 @@
+    "minvalue": 0, "max_deviation": 3, "operation":"-"
+  },
+  "statmelee": {
+    "value": 2"operation":"+"
```
colon in the wrong spot

SandytheLizard replied:
already fixed that

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update amnesia.json

bitcraft commented on tuxemon/technique.py:
```diff
@@ -297,25 +300,38 @@ def changestats(self, user: Monster, target: Monster) -> EffectResult:
+                "/": operator.floordiv,
+                }
+
+                newstatvalue = ops_dict[operation](basestatvalue, value)
```
need to try/except for divide by zero errors.  though unlikely, if the value is zero, and operation is 0, there will be a crash here

SandytheLizard replied:
i dont think thatll happen because the condition above it says to only happen if value >0

bitcraft commented on tuxemon/technique.py:
```diff
@@ -214,7 +215,9 @@ def full_recharge(self) -> None:
         self._combat_counter = 0
-
+    def keep_old_stats(self):
+        self.old_stats_data.append([mon.speed, mon.hp, mon.armour, mon.melee, mon.ranged, mon.dodge])
```
this won't work because `mon` is not in this scope.  please playtest before asking for reviews.

SandytheLizard replied:
fixed

SandytheLizard replied:
ok i added a line to make mon work again

bitcraft replied:
i don't understand.  where is `mon` in this function?

SandytheLizard replied:
sorry, for whatever reason I typed the new line in but it didnt save the changes.

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update technique.py

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr "{user} Overfed {target}."
+
+msgid " dodge_fell"
+msgstr "{user}'s dodge fell. 
```
missing the closing `"`

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr "{user}'s dodge fell. 
+
+msgid "attacks_rose"
+msgstr {user}'s attack damage types rose."
```
missing opening `"`

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
 msgstr "{user} is leeching health from {target}."
 
+msgid "dodge_increased"
+msgstr {user}'s {stat} increased."
```
missing`"`

bitcraft replied:
still missing

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr {user}'s {stat} increased."
+
+msgid "speed_halved"
+msgstr {user}'s speed was halved."
```
missing `"`

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
+msgstr {user}'s speed was halved."
+
+msgid "hp_restored"
+msgstr {user}'s hp was restored."
```
missing `"`

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -169,10 +169,29 @@ msgstr "{name} took poison damage!"
 msgid "combat_state_overfeed_get"
+msgstr "{user} Overfed {target}."
+
+msgid " dodge_fell"
```
extra space before "dodge"

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update base.po

SandytheLizard committed:
Update base.po

SandytheLizard committed:
Update base.po
